Add markup as designed (#574)
unicode-org/message-format-wg: spec/syntax.md

and what outputs might result.
See [function registry](./) for more information.

A _function_ starts with a prefix sigil `:` followed by an _identifier_.

> The _name_ MAY be followed by one or more _options_.

Reviewer: name => identifier

Author: ```suggestion
The _identifier_ MAY be followed by one or more _options_.
```

              / %xE000-10FFFF
```

## Markup

> **_<dfn>Markup</dfn>_** _placeholders_ are _pattern_ parts
> that can be used to represent non-language parts of a _message_,
> such as inline elements or styling that should apply to a span of parts.

Reviewer: I'm not sure about the language here. Perhaps:

```suggestion
**_<dfn>Markup</dfn>_** are _placeholders_ in a _pattern_ that
represent externally defined, non-formattable parts of a _message_.
Unless otherwise provisioned, _markup_ is omitted from the formatted _message_.
Examples of markup might include inline styling or translator
instructions to be applied to a span of _message_ parts.
```

Author: This seems to presume that "formatting" is a process that always produces a string. I don't think that's an accurate portrayal for this spec, which is supporting non-string formatting targets even if it does not explicitly define them.

Reviewer: No, that's a jump. It doesn't say anything about string resolution--it just says that the markup can be omitted ("unless otherwise provisioned"). It doesn't say what form any of that takes. We could expand on the examples, I suppose.

Author: It's still referring to markup as "non-formattable", and states that "_markup_ is omitted [unless] otherwise provisioned". Our design doc, on the other hand, includes this: https://github.com/unicode-org/message-format-wg/blob/2a69d2a17ee7476282bfa824fd4ca62e0a03f6e1/exploration/open-close-placeholders.md?plain=1#L253-L259

So they are formattable, and are not omitted by default _except when formatting to a string_.

Also, we did end up includes standalone, for which placeholders like `{#img/}` or `{#button/}` are rather likely to end up as concrete parts of the output

Reviewer: You're right, but then... how are `img` or `button` different from `b`?

I think what we might be trying to say is:

```suggestion
**_<dfn>Markup</dfn>_** is an implementation-specific
(but otherwise unsupported) mechanism intended to provide for
the inclusion of styling, annotation, or various markup syntaxes
in a _message_ in ways that are compatible with translation.
A _markup_ _placeholder_ can only appear in a _pattern_
and might be omitted or remain unprocessed during the formatting process.
```

Author: > You're right, but then... how are `img` or `button` different from `b`?

They don't have a span of parts within the message that they're wrapping. They explicitly enable the representation of inline elements that should be called out in the explanation of _markup_.

> **_Markup_** is an implementation-specific (but otherwise unsupported) mechanism [...]

What does this mean, and why does it need to be called out as the first thing about markup? How is markup "unsupported"?

My understanding of the MF2 position on _markup_ is that we support syntax for it, and define how to represent it in non-string output. I don't really see how it could be more "supported" than that? Defining the output more precisely would've been much easier if we could have included formatted parts in the spec, but alas, that was not to be.

Maybe @mihnita already has some idea how the ICU4J implementation will handle markup, to give us another point of view?

Reviewer: I wrote the above hurriedly yesterday, regretting "unsupported" even as I suggested it. I don't think it's necessarily "wrong" though. 

The problem with markup is that we don't provide real support for it. It doesn't use functions per-se and we don't provide anything in the registry for it. Open/close don't mean anything, other than as semantic constructs--we don't require them to be in order or paired.

What we're doing is reserving syntax for a partially specified feature. I think there are three ways that implementations will respond to this: (i) they won't support markup; (ii) they'll provide "pass through" support for it using the data model and/or some other means; or (iii) they'll implement either specific markup support (such as for HTML) or some form of pluggable (generic) support.

All implementations can check the syntax of a placeholder and can resolve any variables. Therefore all implementations MUST emit a syntax error for a malformed placeholder and MUST emit a resolution error if they encounter one.

Let's consider an implementation that supports markup. I think we should provide support for implementers who want to put additional requirements on (specific) markup, such as requiring open/close pairing, requiring open/close to be in order, require specific options (or validate option values). Users will appreciate it if a typo in their markup produces an appropriate error (@stasm has cited this in the past). This suggests that we provide an *optional* MarkupError that implementations MAY emit.

What about pass through? I think this is what we currently describe, albeit we require `formatToString` to remove the markup (pass through is to the data model or parts. Here we might want to relax the requirement to allow implementations to decide how to pass markup through (callers might need to reparse the output as a result).

What about non-supporting implementations? They might still emit data model parts or they might eat them. I think we might want to require that they not fail on markup but permit them to not emit the markup.

A key consideration is not to allow markup to become a security attack vector. We don't want `<scr{#bad}ipt>` to become a script tag (or other abusive sequence) in ways that are hard to detect. This can happen if the markup doesn't collapse to nothing until late.

The last two considered together might look like:

- An implementation SHOULD NOT emit the markup placeholder when formatting to string.
- An implementation MAY (SHOULD?) emit unrecognized markup placeholders in the data model (i.e. they do something other than formatToString)
- An implementation SHOULD NOT apply any additional validation (beyond syntax and resolution) to unrecognized or unsupported markup.
- An implementation MAY emit a MarkupError for unrecognized or unsupported markup. (this has to be allowed if implementations that support markup are to identify typos)

Reviewer: I feel strongly about this thread, but think we need to merge the overall changes into the spec. I am filing a separate issue to track this conversation.

Markup comes in three forms:

> **_<dfn>Markup-open</dfn>_** starts with `#` and

Reviewer: Let's start to be fastidious about character references

```suggestion
**_<dfn>Markup-open</dfn>_** starts with U+0023 NUMBER SIGN `#` and
```

> is a _pattern_ part starting a span.

Reviewer: This is slightly misleading, as we don't guarantee closure.

```suggestion
represents an opening element within the _message_,
such as markup used to start a span.
```

> It MAY include _options_, but these are not required.

Reviewer 2: The `but...` is not needed here?

```suggestion
It MAY include _options_.
```

> If a _markup-open_ includes a `/` immediately before its closing `}`,
> it is a self-closing **_<dfn>markup-standalone</dfn>_** part
> representing a standalone element within a message,
> such as an inline image or any other opaque content.

Reviewer: This should be the same as the other two types.

```suggestion
**_<dfn>Markup-standalone</dfn>_** starts with U+0023 NUMBER SIGN `#`
and has a U+002F SOLIDUS `/` immediately before its closing `}`
representing a self-closing or standalone element within the _message_.
It MAY include _options_, but these are not required.
```

> **_<dfn>Markup-close</dfn>_** starts with `/` and

Reviewer: ```suggestion
**_<dfn>Markup-close</dfn>_** starts with U+002F SOLIDUS `/` and
```

is a _pattern_ part ending a span.
Unlike the other forms, it does not include _options_.

```abnf
markup       = "{" [s] markup-open [s] ["/"] "}"
             / "{" [s] markup-close [s] "}"
markup-open  = "#" identifier *(s option)
markup-close = "/" identifier
```

> A _message_ with one `button` markup span and a standalone `img` markup element.
>
> ```
> {#button}Submit{/button} or {#img alt=|Cancel| /}.}
> ```

> _Markup-open_ and _markup-close_ parts SHOULD be paired,

Reviewer 2: Isn't that more of a `MAY`? There exist completely valid systems where markup is stack-based (e.g. TTY) and where I could imagine only `markup-open` being used with success.

Author: ```suggestion
_Markup-open_ and _markup-close_ parts MAY be paired,
```

              / %xE000-10FFFF
```

## Markup

**_<dfn>Markup</dfn>_** _placeholders_ are _pattern_ parts
that can be used to represent non-language parts of a _message_,
such as inline elements or styling that should apply to a span of parts.
Markup comes in three forms:

**_<dfn>Markup-open</dfn>_** starts with U+0023 NUMBER SIGN `#` and
represents an opening element within the _message_,
such as markup used to start a span.
It MAY include _options_.

**_<dfn>Markup-standalone</dfn>_** starts with U+0023 NUMBER SIGN `#`
and has a U+002F SOLIDUS `/` immediately before its closing `}`
representing a self-closing or standalone element within the _message_.
It MAY include _options_.

**_<dfn>Markup-close</dfn>_** starts with U+002F SOLIDUS `/` and
is a _pattern_ part ending a span.
Unlike the other forms, it does not include _options_.

```abnf
markup       = "{" [s] markup-open [s] ["/"] "}"
             / "{" [s] markup-close [s] "}"
markup-open  = "#" identifier *(s option)
markup-close = "/" identifier
```

> A _message_ with one `button` markup span and a standalone `img` markup element.
>
> ```
> {#button}Submit{/button} or {#img alt=|Cancel| /}.}
> ```

> _Markup-open_ and _markup-close_ parts MAY be paired,
> but this is not a requirement.
> A _markup-open_ MAY be present in a message without
> a corresponding _markup-close_, and vice versa.
> _Markup-standalone_ parts are not paired.

Author: @macchiati @aphillips Would this work?

```suggestion
No validity requirements are imposed for the proper pairing or nesting of _markup_.
A _markup-open_ MAY be present in a message without
a corresponding _markup-close_, and vice versa.
```

Reviewer: I noodled with this for a while. I think we want to allow markup regimes to impose requirements while saying that the message is otherwise "valid".

```suggestion
A _markup-open_ can appear without a corresponding _markup-close_.
A _markup-close_ can appear without a corresponding _markup-open_.
_Markup_ _placeholders_ can appear in any order without making the _message_ invalid.
However, specifications or implementations defining _markup_ might impose requirements
on the pairing, ordering, or contents of _markup_ during _formatting_.
```
